cpp/src/strings/substring.cu

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/strings/substring.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/utilities/traits.hpp>
+#include "./utilities.hpp"
+#include "./utilities.cuh"
+
+
+namespace
+{
+
+/**
+ * @brief Used as template parameter to divide size calculation from
+ * the actual string operation within a function.
+ * Useful when most of the logic is identical for both passes.
+ */
+enum TwoPass
```

Comment:
If there are only two options, why not just a clearly named bool template parameter?

Reply:
I like seeing `SizeOnly` and `ExecuteOp` in the code inline. I think it makes the code clearer to see these versus knowing that `!SizeOnly` is actually execution or that `!ExecuteOp` means doing a size-only pass.